rust/main/chains/hyperlane-aleo/src/indexer/traits.rs

```diff
@@ -0,0 +1,226 @@
+use std::collections::HashMap;
+use std::{ops::RangeInclusive, str::FromStr};
+
+use aleo_serialize::AleoSerialize;
+use futures::future;
+use hyperlane_core::{ChainResult, Indexed, LogMeta, H512};
+use snarkvm::prelude::{CanaryV0, Itertools, MainnetV0, Network, TestnetV0};
+use snarkvm::{
+    prelude::{Identifier, Plaintext, ProgramID},
+    synthesizer::program::FinalizeOperation,
+};
+
+use crate::utils::{get_tx_id, to_h256, to_key_id};
+use crate::{AleoProvider, HyperlaneAleoError};
+
+pub(crate) trait AleoIndexer {
+    const INDEX_MAPPING: &str;
+    const VALUE_MAPPING: &str;
+    type AleoType: AleoSerialize<TestnetV0>
+        + AleoSerialize<MainnetV0>
+        + AleoSerialize<CanaryV0>
+        + Into<Self::Type>;
+    type Type;
+
+    fn get_client(&self) -> &AleoProvider;
+
+    fn get_program(&self) -> &str;
+
+    async fn get_finalized_block_number(&self) -> ChainResult<u32> {
+        self.get_client().get_latest_height().await
+    }
+
+    /// Returns the mapping key to index into the event mapping
+    /// For simple u32 indexed mappings, this is just the index as a plaintext
+    /// For hook indexed mappings, this is a struct with hook address and index
+    fn get_mapping_key<N: Network>(&self, index: u32) -> ChainResult<Plaintext<N>> {
+        Ok(index.to_plaintext().map_err(HyperlaneAleoError::from)?)
+    }
```

Comment:
It looks like this implementation is shared only with despatch and delivery indexers. Would it be worth to have this implementation referred only from those indexers?